File: go/validation.go
package zymeup

type ValidationClient struct {
	client *Client
}

func NewValidationClient(c *Client) *ValidationClient {
	return &ValidationClient{client: c}
}

type PhoneVerifyResult struct {
	Valid        bool   `json:"valid"`
	Formatted    string `json:"formatted"`
	CountryCode  string `json:"country_code"`
}

type PhoneFormatResult struct {
	Formatted string `json:"formatted"`
}

type PostalCodeResult struct {
	Valid   bool   `json:"valid"`
	Message string `json:"message"`
	Source  string `json:"source"`
}

type EmailValidationResult struct {
	Valid     bool   `json:"valid"`
	Status    string `json:"status"`
	Message   string `json:"message"`
	Source    string `json:"source"`
	Formatted string `json:"formatted"`
}

type TaxIdValidationResult struct {
	Valid   bool   `json:"valid"`
	Message string `json:"message"`
	Source  string `json:"source"`
}

func (c *ValidationClient) VerifyPhone(countryCode, phone string) (*PhoneVerifyResult, error) {
	path := "/api/v1/validation/phone"
	resp, err := c.client.doRequest("POST", path, map[string]interface{}{
		"country_code": countryCode,
		"phone":        phone,
	})
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                  `json:"code"`
		Data *PhoneVerifyResult  `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ValidationClient) FormatPhone(countryCode, phone string) (*PhoneFormatResult, error) {
	path := "/api/v1/validation/phone/format"
	resp, err := c.client.doRequest("POST", path, map[string]interface{}{
		"country_code": countryCode,
		"phone":        phone,
	})
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                    `json:"code"`
		Data *PhoneFormatResult    `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ValidationClient) ValidatePostalCode(countryCode, code string) (*PostalCodeResult, error) {
	path := "/api/v1/validation/postal-code"
	resp, err := c.client.doRequest("POST", path, map[string]interface{}{
		"country_code": countryCode,
		"code":         code,
	})
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                    `json:"code"`
		Data *PostalCodeResult     `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ValidationClient) ValidateEmail(email string) (*EmailValidationResult, error) {
	path := "/api/v1/validation/email"
	resp, err := c.client.doRequest("POST", path, map[string]interface{}{
		"email": email,
	})
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                        `json:"code"`
		Data *EmailValidationResult    `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (c *ValidationClient) ValidateTaxId(countryCode, taxId string) (*TaxIdValidationResult, error) {
	path := "/api/v1/validation/tax-id"
	resp, err := c.client.doRequest("POST", path, map[string]interface{}{
		"country_code": countryCode,
		"tax_id":       taxId,
	})
	if err != nil {
		return nil, err
	}
	var result struct {
		Code int                      `json:"code"`
		Data *TaxIdValidationResult  `json:"data"`
	}
	if err := decodeResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

File: rust/src/validation.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;

#[derive(Debug, serde::Deserialize)]
pub struct PhoneVerifyResult {
    pub valid: bool,
    #[serde(rename = "formatted")]
    pub formatted: String,
    #[serde(rename = "country_code")]
    pub country_code: String,
}

#[derive(Debug, serde::Deserialize)]
pub struct PhoneFormatResult {
    #[serde(rename = "formatted")]
    pub formatted: String,
}

#[derive(Debug, serde::Deserialize)]
pub struct PostalCodeResult {
    pub valid: bool,
    pub message: String,
    pub source: String,
}

#[derive(Debug, serde::Deserialize)]
pub struct EmailValidationResult {
    pub valid: bool,
    pub status: String,
    pub message: String,
    pub source: String,
    #[serde(rename = "formatted")]
    pub formatted: String,
}

#[derive(Debug, serde::Deserialize)]
pub struct TaxIdValidationResult {
    pub valid: bool,
    pub message: String,
    pub source: String,
}

pub struct ValidationClient {
    pub(crate) inner: HttpClient,
}

impl ValidationClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }

    pub async fn verify_phone(&self, country_code: &str, phone: &str) -> Result<PhoneVerifyResult> {
        self.inner
            .request(
                "/api/v1/validation/phone",
                Method::POST,
                Some(serde_json::json!({
                    "country_code": country_code,
                    "phone": phone,
                })),
            )
            .await
    }

    pub async fn format_phone(&self, country_code: &str, phone: &str) -> Result<PhoneFormatResult> {
        self.inner
            .request(
                "/api/v1/validation/phone/format",
                Method::POST,
                Some(serde_json::json!({
                    "country_code": country_code,
                    "phone": phone,
                })),
            )
            .await
    }

    pub async fn validate_postal_code(&self, country_code: &str, code: &str) -> Result<PostalCodeResult> {
        self.inner
            .request(
                "/api/v1/validation/postal-code",
                Method::POST,
                Some(serde_json::json!({
                    "country_code": country_code,
                    "code": code,
                })),
            )
            .await
    }

    pub async fn validate_email(&self, email: &str) -> Result<EmailValidationResult> {
        self.inner
            .request(
                "/api/v1/validation/email",
                Method::POST,
                Some(serde_json::json!({
                    "email": email,
                })),
            )
            .await
    }

    pub async fn validate_tax_id(&self, country_code: &str, tax_id: &str) -> Result<TaxIdValidationResult> {
        self.inner
            .request(
                "/api/v1/validation/tax-id",
                Method::POST,
                Some(serde_json::json!({
                    "country_code": country_code,
                    "tax_id": tax_id,
                })),
            )
            .await
    }
}
